colossalai/shardformer/layer/loss.py

```diff
@@ -226,13 +225,13 @@ def dist_cross_entropy(
             logits,
             labels,
             process_group=shard_config.tensor_parallel_process_group,
-            vocab_size=out_features,
+            vocab_size=vocab_size,
```

Comment:
As we may pad lm head, vocab size is not always equal to out features. I doubt whether we can use only one argument here.

Reply:
In my experience, this works for all models currently in shardformer. And we're actually passing in `lm_head.out_features`?

Reply:
@flybird11111 Can you take a look?

Reply:
> In my experience, this works for all models currently in shardformer, since we're actually passing in `lm_head.out_features`?

Hi, lm.out_features is the filled vocabulary size. When the weight of lm_head is filled, it is not equal to config.vocabe_size at all. The vocabulary_size of the test_kit model can be divisible by 64 and does not require padding, so there is no problem.

Reply:
In any case, this won't change the loss since no tokens will be mapped to padded ids?

Reply:
> In any case, this won't change the loss since no tokens will be mapped to padded ids?

The padded vocabulary size is used to calculate the vocabulary size per partition. If the last rank uses logit.size[-1] for the vocabulary size per partition, it will result in an error.

Reply:
I think the padded tokens won't contribute to the loss. I've tried chaging llama's vocab size to 32001, and the tests worked fine.

Reply:
> In any case, this won't change the loss since no tokens will be mapped to padded ids?

Because the output of the last rank is the unpadded size.
<img width="591" alt="image" src="https://github.com/user-attachments/assets/ad2a4204-e258-4e90-bd34-6b6837f163ae">

Reply:
Sorry, you provided the correct padded vocabulary size. It's okay.